### ISSUE TITLE
Dataset operations: returned type

### ISSUE THREAD

**Issue by NicoLaval** [FranckCo/VTL-Community#3]
## Use case description

[Lunatic](https://github.com/InseeFr/Lunatic) is a React library allowing questionnaire instantiation and management.
In order to vary labels, calculate variables and boost filters, Lunatic call [Trevas JS](https://github.com/InseeFr/Trevas-JS), a VTL 2.0 JavaScript engine.

We are currently beginning to support Dataset operations, but are facing issues when we try to access output bindings.

Considering SUM operator and DS defined as below:

|id|me_1|
|:-:|:-:|
|"1"|1|
|"2"|2|
|"1"|3|

Regarding to VTL specification, we expect to get following data in the result dataset:
- `sum(DS)` --> me_1 = [6]
- `sum(DS group by id)` --> id = ["1", "2"] & me_1 = [4, 2]

In the first case, I'm only interested by the 6 value (as a scalar), and in the second, by the full vector.

Nevertheless, by design, VTL doesn't provide getter.
To implement this type of operator completely, sometimes wishing to access the single scalar value of a component, and other times the entire component we don't find a way to tell the engine to reduce the dimension of the result or not.

## Current DS model into [Trevas JS](https://github.com/InseeFr/Trevas-JS)

```json
{
    "dataStructure": {
        "id": { role: "IDENTIFIER", type: "STRING" },
        "me_1": { role: "MEASURE", type: "INTEGER" },
    },
    "dataPoints": {
        "id": ["a", "b", "c"],
        "me_1": [1, 2, 3],
    }
}
```

## Proposal

Adding a resultType (with an ENUM range: "SCALAR" or "COMPONENT") to the Dataset model will allow to get directly the sum of my measure:

```json
{
    "resultType": "SCALAR"
    "dataStructure": {
        "id": { role: "IDENTIFIER", type: "STRING" },
        "me_1": { role: "MEASURE", type: "INTEGER" },
    },
    "dataPoints": {
        "id": ["a", "b", "c"],
        "me_1": [1, 2, 3],
    }
}
```

**Comment by vpinna80** [comment]
Combining sum and sub? `sum(DS[sub id = "1"])` will get the single scalar value.

That dataset is malformed though. Ids cannot repeat in VTL.

**Comment by NicoLaval** [comment]
`sub` is interesting, but if we only sum, without group by, ids will disappear from result, and it will be impossible to use `sub` to select result. That's right ?

Yes, example was bad, with a second measure it's fixed:

|id|id2|me_1|
|:-:|:-:|:-:|
|"1"|"a"|1|
|"2"|"a"|2|
|"1"|"b"|3|

**Comment by vpinna80** [comment]
Aren't these enough?

`DS[sub id = "1", id2 = "a"]` --> 2 as scalar value
`sum(DS[sub id = "1")` --> 4 as scalar value
`DS[sub id = "1"]` --> a dataset:
|id2|me_1|
|:-:|:-:|
|"a"|1|
|"b"|3|

**Comment by NicoLaval** [comment]
`DS[sub id = "1", id2 = "a"]` --> 1 as a scalar I think

I agree with other examples, but if I want to sum all values of me_1?

I just imagine `sum(DS)` which will return a dataset:

|me_1|
|:-:|
|6|

But not 6 as a scalar value.

**Comment by vpinna80** [comment]
Ah yes, inside VTL there would be only minimal semantic differences between a scalar value and a mono-measured dataset with no identifiers.
Outside of it, another story.
Can a global configuration be used perhaps? Like "PreferScalarResult=true"

**Comment by NicoLaval** [comment]
Where would you defined it?

Into Dataset Model directly? (just replacing the `resultType` in my proposal?)

**Comment by vpinna80** [comment]
I'll think about it, but I'd prefer to have this on a global setting rather than on each dataset.
After all, you need this only in specific computations and it'd be completely pointless if you have more than 1 measure.

**Comment by NicoLaval** [comment]
It can be useful even if I want to sum several measures over the entire dataset. That's why I guessed it would be right to introduce an option for each call instead of a global VTL rule.

**Comment by antonio-olleros** [comment]
I think this is a very relevant question, but I disagree with @vpinna80, because I think there is a big difference between a monomeasure dataset without identifiers and a scalar; notably, you cannot use a dataset inside a calc (this is a real use case we have: We do some calculations to get a scalar that we want to then add as a fixed value for a new component with calc). In our opinion, at the moment, we cannot do this, if we are strict with the VTL specification, and we need to be able to do this. In the pass, I proposed to add a new operator to extract a scalar from datasets (for instance ##), but for me any other option would work as long it is shared and explicit

**Comment by vpinna80** [comment]
Actually this was one of the first items discussed at the start of the development of our engine.
Unfortunately, the reference manual says that the calc/aggr expression can only have components that are in the input dataset (6802, 6821).

**Comment by antonio-olleros** [comment]
Yes! And if we include in a join a dataset without identifiers, there is never going to be a match, by definition. The only thing that can be done to add a scalar result to another dataset is to do a cross join, which is not nice at all...

**Comment by hadrienk** [comment]
What about implicit casting between scalar/dataset under certain conditions?

**Comment by antonio-olleros** [comment]
I think that this can be a good solution. I think that it would be enough with having this implicit casting when using a dataset monomeasure and without identifiers inside a calc; at least that would cover all the use cases we have had.
